#!/bin/bash
# After synthetic reads are generated by RNFtools, run the mapping tool with 
# different options (best-match and report-all)

# Build index out of reference genome
bowtie2-build /home/mohammad/datasets/bacteria/Orientia_tsutsugamushi_Ikeda_uid58869/NC_010793.fna ot

# Run Bowtie2 for mapping reads

# Paired mapping, best-match
bowtie2 -x ot -X 650 -1 ot_reads.1.fq -2 ot_reads.2.fq -S ot-paired-mapping-best-match.sam 2> log-ot-paired-mapping-best-match.txt
samtools view -bS ot-paired-mapping-best-match.sam > ot-paired-mapping-best-match.bam

# Paired mapping, report-all
bowtie2 -a -x ot -X 650 -1 ot_reads.1.fq -2 ot_reads.2.fq -S ot-paired-mapping-report-all.sam 2> log-ot-paired-mapping-report-all.txt
samtools view -bS ot-paired-mapping-report-all.sam > ot-paired-mapping-report-all.bam

# Single mapping, best-match
bowtie2 -x ot -X 650 -U ot_reads.1.fq,ot_reads.2.fq -S ot-single-mapping-best-match.sam 2> log-ot-single-mapping-best-match.txt
samtools view -bS ot-single-mapping-best-match.sam > ot-single-mapping-best-match.bam

# Single mapping, report-all
bowtie2 -a -x ot -X 650 -U ot_reads.1.fq,ot_reads.2.fq -S ot-single-mapping-report-all.sam 2> log-ot-single-mapping-best-match.txt
samtools view -bS ot-single-mapping-report-all.sam > ot-single-mapping-report-all.bam

echo "\n=== Done ===\n"
